sdk/logical/storage_view.go

```diff
@@ -9,37 +9,80 @@ import (
 	"strings"
 )
 
-type StorageView struct {
-	storage Storage
-	prefix  string
+type StorageViewCore interface {
+	Prefix() string
+	SubView(prefix string) StorageView
+	SanityCheck(key string) error
+	ExpandKey(suffix string) string
+	TruncateKey(full string) string
+}
+
+type StorageView interface {
+	Storage
+	StorageViewCore
+}
+
+type TransactionalStorageView interface {
+	TransactionalStorage
+	StorageViewCore
+}
+
+// Note that, within a transaction, creating and committing from a
+// SubView commits the entire transaction.
+type StorageViewTransaction interface {
+	Transaction
+	StorageViewCore
 }
 
 var ErrRelativePath = errors.New("relative paths not supported")
 
-func NewStorageView(storage Storage, prefix string) *StorageView {
-	return &StorageView{
+func NewStorageView(storage Storage, prefix string) StorageView {
+	sv := &storageView{
 		storage: storage,
 		prefix:  prefix,
 	}
+
+	if _, ok := storage.(TransactionalStorage); ok {
+		return &transactionalStorageView{
+			*sv,
+		}
+	}
+
+	if _, ok := storage.(Transaction); ok {
+		return &storageViewTransaction{
+			*sv,
+		}
+	}
+
+	return sv
+}
+
+type storageView struct {
+	storage Storage
+	prefix  string
 }
 
+var (
+	_ Storage     = &storageView{}
+	_ StorageView = &storageView{}
+)
+
 // logical.Storage impl.
-func (s *StorageView) List(ctx context.Context, prefix string) ([]string, error) {
+func (s *storageView) List(ctx context.Context, prefix string) ([]string, error) {
```

Comment:
@cipherboy : is this a kind of Go encapsulation mechanism ? The public method is defined with a private (package) receiver.

Reply:
Not really.

Go has the concept of an interface, which is an arbitrary grouping of methods. You can type assert between various interfaces or implementations there-of (so `var x logical.StorageView` could be type asserted to `x.(*storageView)` if you were in `logical`. 

The interface itself (capital S `StorageView`) cannot have default implementations attached to it (unless they added this and I missed it).

At least in this code base, and in a few others I've seen, it is relatively common to hide the implementation of a public interface behind a private struct implementing it.

So, by making the rename change above (`type StorageView struct {}` --> `type storageView struct {}` and adding `type StorageView interface{}` in its place), we get a way of hiding the implementation details behind an interface (collection of methods) and can update the only callers that bothered to use it directly (in this case, `vault/barrier_view.go`).

Reply:
@cipherboy : Yes, thank you, that's what I thought, but I wasn't sure: it allows to hide implementation details for a given interface.